onnxruntime/core/providers/cpu/tensor/gather_nd.cc

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "gather_nd.h"
+
+namespace onnxruntime {
+
+// Register a kernel for kMsDomain (contrib op) GatherND
+#ifndef DISABLE_CONTRIB_OPS
+
+namespace contrib {
+// TODO: Remove this contrib kernel registration and the schema from the appropriate places
+// once Keras Mask RCNN is shipped with all ONNX domain ops
+
+// Currently this kernel is required to support Keras Mask-RCNN
+ONNX_OPERATOR_KERNEL_EX(GatherND, kMSDomain, 1, kCpuExecutionProvider,
+                        KernelDefBuilder()
+                            .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+                            // contrib spec supports `int32_t` and `int64_t` for indices
+                            .TypeConstraint("Tind", {DataTypeImpl::GetTensorType<int32_t>(),
+                                                     DataTypeImpl::GetTensorType<int64_t>()}),
+                        GatherND);
+
+}  // namespace contrib
+
+#endif
+
+ONNX_CPU_OPERATOR_KERNEL(GatherND, 11,
+                         KernelDefBuilder()
+                             .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+                             // official ONNX spec only supports `int64_t` for indices
+                             .TypeConstraint("Tind", DataTypeImpl::GetTensorType<int64_t>()),
+                         GatherND);
+
+template <typename Tind>
+Status GatherNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p) const {
+  auto input_tensor = context->Input<Tensor>(0);
+  auto indice_tensor = context->Input<Tensor>(1);
+  ORT_ENFORCE(input_tensor != nullptr);
+  ORT_ENFORCE(indice_tensor != nullptr);
+
+  auto input_shape = input_tensor->Shape();
+  auto indice_shape = indice_tensor->Shape();
+  if (indice_shape.NumDimensions() == 0) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "indices tensor must has rank larger than 0");
+  }
+
+  auto last_indice_dimension = indice_shape[indice_shape.NumDimensions() - 1];
+  if (last_indice_dimension > static_cast<int64_t>(input_shape.NumDimensions())) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "last dimension of indices must not be larger than rank of input tensor");
+  }
+
+  std::vector<int64_t> shape(indice_shape.GetDims().begin(), indice_shape.GetDims().end() - 1);
+  shape.insert(shape.end(), input_shape.GetDims().begin() + last_indice_dimension, input_shape.GetDims().end());
+  auto output_tensor = context->Output(0, TensorShape(shape));
+  std::vector<int64_t> element_counts(last_indice_dimension,
```

Comment:
std::move(shape)

Reply:
I believe we don't really need this. Since the ctor takes a reference anyway ?